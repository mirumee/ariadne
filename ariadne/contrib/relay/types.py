from collections import namedtuple
from typing import Any, Callable, Dict

from typing_extensions import TypeVar

from ariadne.contrib.relay.connection import RelayConnection

ConnectionResolver = TypeVar("ConnectionResolver", bound=Callable[..., RelayConnection])
GlobalIDTuple = namedtuple("GlobalIDTuple", ["type", "id"])
GlobalIDDecoder = Callable[[Dict[str, Any]], GlobalIDTuple]
