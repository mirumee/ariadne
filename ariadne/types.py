from inspect import isawaitable
from dataclasses import dataclass
from typing import (
    Any,
    AsyncGenerator,
    Callable,
    Collection,
    List,
    Optional,
    Tuple,
    Type,
    Union,
    TypeVar,
)
from typing_extensions import Protocol, runtime_checkable

from graphql import (
    DocumentNode,
    ExecutionResult,
    GraphQLError,
    GraphQLResolveInfo,
    GraphQLSchema,
)
from graphql.validation.rules import ASTValidationRule
from graphql.execution import Middleware

from starlette.websockets import WebSocket


# Note: this should be [Any, GraphQLResolveInfo, **kwargs],
# but this is not achieveable with python types yet:
# https://github.com/mirumee/ariadne/pull/79
Resolver = Callable[..., Any]
GraphQLResult = Tuple[bool, dict]
SubscriptionResult = Tuple[
    bool, Union[List[dict], AsyncGenerator[ExecutionResult, None]]
]
Subscriber = Callable[..., AsyncGenerator]
ErrorFormatter = Callable[[GraphQLError, bool], dict]

ContextValue = Union[Any, Callable[[Any], Any]]
RootValue = Union[Any, Callable[[Optional[Any], DocumentNode], Any]]

ValidationRules = Union[
    Collection[Type[ASTValidationRule]],
    Callable[
        [Optional[Any], DocumentNode, dict],
        Optional[Collection[Type[ASTValidationRule]]],
    ],
]

ExtensionList = Optional[List[Union[Type["Extension"], Callable[[], "Extension"]]]]


Extensions = Union[
    Callable[[Any, Optional[ContextValue]], ExtensionList], ExtensionList
]
MiddlewareList = Optional[List[Middleware]]
Middlewares = Union[
    Callable[[Any, Optional[ContextValue]], MiddlewareList], MiddlewareList
]


@dataclass
class Operation:
    id: str
    name: Optional[str]
    generator: AsyncGenerator


OnConnect = Callable[[WebSocket, Any], Any]
OnDisconnect = Callable[[WebSocket], Any]
OnOperation = Callable[[WebSocket, Operation], Any]
OnComplete = Callable[[WebSocket, Operation], Any]


class WebSocketConnectionError(Exception):
    """Special error class enabling custom error reporting for on_connect"""

    def __init__(self, payload: Optional[Union[dict, str]] = None):
        if isinstance(payload, dict):
            self.payload = payload
        elif payload:
            self.payload = {"message": str(payload)}
        else:
            self.payload = {"message": "Unexpected error has occurred."}


@runtime_checkable
class Extension(Protocol):
    def request_started(self, context: ContextValue):
        pass  # pragma: no cover

    def request_finished(self, context: ContextValue):
        pass  # pragma: no cover

    async def resolve(
        self, next_: Resolver, obj: Any, info: GraphQLResolveInfo, **kwargs
    ):
        result = next_(obj, info, **kwargs)
        if isawaitable(result):
            result = await result
        return result

    def has_errors(self, errors: List[GraphQLError], context: ContextValue):
        pass  # pragma: no cover

    def format(self, context: ContextValue) -> Optional[dict]:
        pass  # pragma: no cover


class ExtensionSync(Extension):
    def resolve(
        self, next_: Resolver, obj: Any, info: GraphQLResolveInfo, **kwargs
    ):  # pylint: disable=invalid-overridden-method
        return next_(obj, info, **kwargs)


@runtime_checkable
class SchemaBindable(Protocol):
    def bind_to_schema(self, schema: GraphQLSchema) -> None:
        pass  # pragma: no cover


SubscriptionHandler = TypeVar("SubscriptionHandler")
SubscriptionHandlers = Union[
    Tuple[Type[SubscriptionHandler]], List[Type[SubscriptionHandler]]
]
