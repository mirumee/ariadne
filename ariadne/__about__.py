__version__ = "0.0.1.dev0"  # This is overwritten by Hatch in CI/CD, don't change it.
