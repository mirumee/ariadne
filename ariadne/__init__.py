from .enums import (
    EnumType,
    set_default_enum_values_on_schema,
    validate_schema_enum_values,
)
from .executable_schema import make_executable_schema
from .extensions import ExtensionManager
from .file_uploads import combine_multipart_data, upload_scalar
from .format_error import (
    format_error,
    get_error_extension,
    get_formatted_error_context,
    get_formatted_error_traceback,
)
from .graphql import graphql, graphql_sync, subscribe
from .inputs import InputType
from .interfaces import InterfaceType, type_implements_interface
from .load_schema import load_schema_from_path
from .objects import MutationType, ObjectType, QueryType
from .resolvers import (
    FallbackResolversSetter,
    SnakeCaseFallbackResolversSetter,
    fallback_resolvers,
    is_default_resolver,
    resolve_to,
    snake_case_fallback_resolvers,
)
from .scalars import ScalarType
from .schema_names import SchemaNameConverter, convert_schema_names
from .schema_visitor import SchemaDirectiveVisitor
from .subscriptions import SubscriptionType
from .types import Extension, SchemaBindable
from .unions import UnionType
from .utils import (
    convert_camel_case_to_snake,
    convert_kwargs_to_snake_case,
    gql,
    unwrap_graphql_error,
)

__all__ = [
    "EnumType",
    "Extension",
    "ExtensionManager",
    "FallbackResolversSetter",
    "InputType",
    "InterfaceType",
    "MutationType",
    "ObjectType",
    "QueryType",
    "ScalarType",
    "SchemaBindable",
    "SchemaDirectiveVisitor",
    "SchemaNameConverter",
    "SnakeCaseFallbackResolversSetter",
    "SubscriptionType",
    "UnionType",
    "combine_multipart_data",
    "convert_camel_case_to_snake",
    "convert_kwargs_to_snake_case",
    "convert_schema_names",
    "fallback_resolvers",
    "format_error",
    "get_error_extension",
    "get_formatted_error_context",
    "get_formatted_error_traceback",
    "gql",
    "graphql",
    "graphql_sync",
    "is_default_resolver",
    "load_schema_from_path",
    "make_executable_schema",
    "resolve_to",
    "set_default_enum_values_on_schema",
    "snake_case_fallback_resolvers",
    "subscribe",
    "type_implements_interface",
    "unwrap_graphql_error",
    "upload_scalar",
    "validate_schema_enum_values",
]
