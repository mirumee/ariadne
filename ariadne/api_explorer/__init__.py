from .api_explorer import APIExplorer, APIExplorerHttp405
from .graphiql import APIExplorerGraphiQL
from .playground import APIExplorerPlayground

__all__ = [
    "APIExplorer",
    "APIExplorerGraphiQL",
    "APIExplorerHttp405",
    "APIExplorerPlayground",
]
