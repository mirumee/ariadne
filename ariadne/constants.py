DATA_TYPE_JSON = "application/json"
DATA_TYPE_MULTIPART = "multipart/form-data"

CONTENT_TYPE_JSON = "application/json; charset=UTF-8"
CONTENT_TYPE_TEXT_HTML = "text/html; charset=UTF-8"
CONTENT_TYPE_TEXT_PLAIN = "text/plain; charset=UTF-8"

HTTP_STATUS_200_OK = "200 OK"
HTTP_STATUS_400_BAD_REQUEST = "400 Bad Request"
HTTP_STATUS_405_METHOD_NOT_ALLOWED = "405 Method Not Allowed"
