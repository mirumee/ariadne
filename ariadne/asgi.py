import asyncio
import json
from typing import (
    Any,
    AsyncGenerator,
    Callable,
    Dict,
    List,
    Optional,
    Union,
    Type,
    cast,
)

from graphql import GraphQLError, GraphQLSchema
from graphql.execution import MiddlewareManager
from starlette.datastructures import UploadFile
from starlette.requests import Request
from starlette.responses import HTMLResponse, JSONResponse, PlainTextResponse, Response
from starlette.types import Receive, Scope, Send
from starlette.websockets import WebSocket, WebSocketState, WebSocketDisconnect

from .constants import DATA_TYPE_JSON, DATA_TYPE_MULTIPART, PLAYGROUND_HTML
from .exceptions import HttpBadRequestError, HttpError
from .file_uploads import combine_multipart_data
from .format_error import format_error
from .graphql import graphql, subscribe
from .logger import log_error
from .types import ContextValue, ErrorFormatter, Extension, RootValue

GQL_CONNECTION_INIT = "connection_init"  # Client -> Server
GQL_CONNECTION_ACK = "connection_ack"  # Server -> Client
GQL_CONNECTION_ERROR = "connection_error"  # Server -> Client

# NOTE: The keep alive message type does not follow the standard due to connection optimizations
GQL_CONNECTION_KEEP_ALIVE = "ka"  # Server -> Client

GQL_CONNECTION_TERMINATE = "connection_terminate"  # Client -> Server
GQL_START = "start"  # Client -> Server
GQL_DATA = "data"  # Server -> Client
GQL_ERROR = "error"  # Server -> Client
GQL_COMPLETE = "complete"  # Server -> Client
GQL_STOP = "stop"  # Client -> Server

ExtensionList = Optional[List[Type[Extension]]]


class GraphQL:
    def __init__(
        self,
        schema: GraphQLSchema,
        *,
        context_value: Optional[ContextValue] = None,
        root_value: Optional[RootValue] = None,
        debug: bool = False,
        logger: Optional[str] = None,
        error_formatter: ErrorFormatter = format_error,
        extensions: Union[Callable[[Any], ExtensionList], ExtensionList] = None,
        middleware: Optional[MiddlewareManager] = None,
        keepalive: float = None,
    ):
        self.context_value = context_value
        self.root_value = root_value
        self.debug = debug
        self.logger = logger
        self.error_formatter = error_formatter
        self.extensions = extensions
        self.middleware = middleware
        self.keepalive = keepalive
        self.schema = schema

    async def __call__(self, scope: Scope, receive: Receive, send: Send):
        if scope["type"] == "http":
            await self.handle_http(scope=scope, receive=receive, send=send)
        elif scope["type"] == "websocket":
            await self.handle_websocket(scope=scope, receive=receive, send=send)
        else:
            raise ValueError("Unknown scope type: %r" % (scope["type"],))

    async def get_context_for_request(self, request: Any) -> Any:
        if callable(self.context_value):
            return self.context_value(request)
        return self.context_value or {"request": request}

    async def get_extensions_for_request(self, request: Any) -> ExtensionList:
        if callable(self.extensions):
            return self.extensions(request)
        return self.extensions

    async def handle_http(self, scope: Scope, receive: Receive, send: Send):
        request = Request(scope=scope, receive=receive)
        if request.method == "GET":
            response = await self.render_playground(request)
        elif request.method == "POST":
            response = await self.graphql_http_server(request)
        else:
            response = Response(status_code=405)
        await response(scope, receive, send)

    async def handle_websocket(self, scope: Scope, receive: Receive, send: Send):
        websocket = WebSocket(scope=scope, receive=receive, send=send)
        await self.websocket_server(websocket)

    async def render_playground(  # pylint: disable=unused-argument
        self, request: Request
    ) -> Response:
        return HTMLResponse(PLAYGROUND_HTML)

    async def graphql_http_server(self, request: Request) -> Response:
        try:
            data = await self.extract_data_from_request(request)
        except HttpError as error:
            return PlainTextResponse(error.message or error.status, status_code=400)

        context_value = await self.get_context_for_request(request)
        extensions = await self.get_extensions_for_request(request)

        success, response = await graphql(
            self.schema,
            data,
            context_value=context_value,
            root_value=self.root_value,
            debug=self.debug,
            logger=self.logger,
            error_formatter=self.error_formatter,
            extensions=extensions,
            middleware=self.middleware,
        )
        status_code = 200 if success else 400
        return JSONResponse(response, status_code=status_code)

    async def extract_data_from_request(self, request: Request):
        content_type = request.headers.get("Content-Type", "")
        content_type = content_type.split(";")[0]

        if content_type == DATA_TYPE_JSON:
            return await self.extract_data_from_json_request(request)
        if content_type == DATA_TYPE_MULTIPART:
            return await self.extract_data_from_multipart_request(request)

        raise HttpBadRequestError(
            "Posted content must be of type {} or {}".format(
                DATA_TYPE_JSON, DATA_TYPE_MULTIPART
            )
        )

    async def extract_data_from_json_request(self, request: Request):
        try:
            return await request.json()
        except (TypeError, ValueError):
            raise HttpBadRequestError("Request body is not a valid JSON")

    async def extract_data_from_multipart_request(self, request: Request):
        try:
            request_body = await request.form()
        except ValueError:
            raise HttpBadRequestError("Request body is not a valid multipart/form-data")

        try:
            operations = json.loads(request_body.get("operations"))
        except (TypeError, ValueError):
            raise HttpBadRequestError(
                "Request 'operations' multipart field is not a valid JSON"
            )
        try:
            files_map = json.loads(request_body.get("map"))
        except (TypeError, ValueError):
            raise HttpBadRequestError(
                "Request 'map' multipart field is not a valid JSON"
            )

        request_files = {
            key: value
            for key, value in request_body.items()
            if isinstance(value, UploadFile)
        }

        return combine_multipart_data(operations, files_map, request_files)

    async def websocket_server(self, websocket: WebSocket) -> None:
        subscriptions: Dict[str, AsyncGenerator] = {}
        await websocket.accept("graphql-ws")
        try:
            while websocket.application_state != WebSocketState.DISCONNECTED:
                message = await websocket.receive_json()
                await self.handle_websocket_message(message, websocket, subscriptions)
        except WebSocketDisconnect:
            pass
        finally:
            for operation_id in subscriptions:
                await subscriptions[operation_id].aclose()

    async def handle_websocket_message(
        self,
        message: dict,
        websocket: WebSocket,
        subscriptions: Dict[str, AsyncGenerator],
    ):
        operation_id = cast(str, message.get("id"))
        message_type = cast(str, message.get("type"))

        if message_type == GQL_CONNECTION_INIT:
            await websocket.send_json({"type": GQL_CONNECTION_ACK})
            asyncio.ensure_future(self.keep_websocket_alive(websocket))
        elif message_type == GQL_CONNECTION_TERMINATE:
            await websocket.close()
        elif message_type == GQL_START:
            await self.start_websocket_subscription(
                message.get("payload"), operation_id, websocket, subscriptions
            )
        elif message_type == GQL_STOP:
            if operation_id in subscriptions:
                await subscriptions[operation_id].aclose()
                del subscriptions[operation_id]

    async def keep_websocket_alive(self, websocket: WebSocket):
        if not self.keepalive:
            return
        while websocket.application_state != WebSocketState.DISCONNECTED:
            try:
                await websocket.send_json({"type": GQL_CONNECTION_KEEP_ALIVE})
            except WebSocketDisconnect:
                return
            await asyncio.sleep(self.keepalive)

    async def start_websocket_subscription(
        self,
        data: Any,
        operation_id: str,
        websocket: WebSocket,
        subscriptions: Dict[str, AsyncGenerator],
    ):
        context_value = await self.get_context_for_request(websocket)
        success, results = await subscribe(
            self.schema,
            data,
            context_value=context_value,
            root_value=self.root_value,
            debug=self.debug,
            logger=self.logger,
            error_formatter=self.error_formatter,
        )
        if not success:
            results = cast(List[dict], results)
            await websocket.send_json(
                {"type": GQL_ERROR, "id": operation_id, "payload": results[0]}
            )
        else:
            results = cast(AsyncGenerator, results)
            subscriptions[operation_id] = results
            asyncio.ensure_future(
                self.observe_async_results(results, operation_id, websocket)
            )

    async def observe_async_results(
        self, results: AsyncGenerator, operation_id: str, websocket: WebSocket
    ) -> None:
        try:
            async for result in results:
                payload = {}
                if result.data:
                    payload["data"] = result.data
                if result.errors:
                    for error in result.errors:
                        log_error(error, self.logger)
                    payload["errors"] = [
                        self.error_formatter(error, self.debug)
                        for error in result.errors
                    ]
                await websocket.send_json(
                    {"type": GQL_DATA, "id": operation_id, "payload": payload}
                )
        except Exception as error:
            if not isinstance(error, GraphQLError):
                error = GraphQLError(str(error), original_error=error)
            log_error(error, self.logger)
            payload = {"errors": [self.error_formatter(error, self.debug)]}
            await websocket.send_json(
                {"type": GQL_DATA, "id": operation_id, "payload": payload}
            )

        if websocket.application_state != WebSocketState.DISCONNECTED:
            await websocket.send_json({"type": GQL_COMPLETE, "id": operation_id})
