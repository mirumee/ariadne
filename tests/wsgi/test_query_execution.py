from werkzeug.test import Client
from werkzeug.wrappers import Response

from ariadne.wsgi import GraphQL
from ariadne.constants import HTTP_STATUS_200_OK, HTTP_STATUS_400_BAD_REQUEST
from ariadne.types import ExtensionSync

from .factories import create_multipart_request


# Add TestClient to keep test similar to ASGI
class TestClient(Client):
    __test__ = False

    def __init__(self, app):
        super().__init__(app, Response)


operation_name = "SayHello"
variables = {"name": "Bob"}
complex_query = """
  query SayHello($name: String!) {
    hello(name: $name)
  }
"""


def test_query_is_executed_for_post_json_request(
    middleware,
    start_response,
    graphql_query_request_factory,
    graphql_response_headers,
    assert_json_response_equals_snapshot,
):
    request = graphql_query_request_factory(query="{ status }")
    result = middleware(request, start_response)
    start_response.assert_called_once_with(HTTP_STATUS_200_OK, graphql_response_headers)
    assert_json_response_equals_snapshot(result)


def test_complex_query_is_executed_for_post_json_request(
    middleware,
    start_response,
    graphql_query_request_factory,
    graphql_response_headers,
    assert_json_response_equals_snapshot,
):
    request = graphql_query_request_factory(
        query=complex_query, variables=variables, operationName=operation_name
    )
    result = middleware(request, start_response)
    start_response.assert_called_once_with(HTTP_STATUS_200_OK, graphql_response_headers)
    assert_json_response_equals_snapshot(result)


def test_complex_query_without_operation_name_executes_successfully(
    middleware,
    start_response,
    graphql_query_request_factory,
    graphql_response_headers,
    assert_json_response_equals_snapshot,
):
    request = graphql_query_request_factory(query=complex_query, variables=variables)
    result = middleware(request, start_response)
    start_response.assert_called_once_with(HTTP_STATUS_200_OK, graphql_response_headers)
    assert_json_response_equals_snapshot(result)


def test_attempt_execute_complex_query_without_variables_returns_error_json(
    middleware,
    start_response,
    graphql_query_request_factory,
    graphql_response_headers,
    assert_json_response_equals_snapshot,
):
    request = graphql_query_request_factory(
        query=complex_query, operationName=operation_name
    )
    result = middleware(request, start_response)
    start_response.assert_called_once_with(HTTP_STATUS_200_OK, graphql_response_headers)
    assert_json_response_equals_snapshot(result)


def test_attempt_execute_query_without_query_entry_returns_error_json(
    middleware,
    start_response,
    graphql_query_request_factory,
    graphql_response_headers,
    assert_json_response_equals_snapshot,
):
    request = graphql_query_request_factory(variables=variables)
    result = middleware(request, start_response)
    start_response.assert_called_once_with(
        HTTP_STATUS_400_BAD_REQUEST, graphql_response_headers
    )
    assert_json_response_equals_snapshot(result)


def test_attempt_execute_query_with_non_string_query_returns_error_json(
    middleware,
    start_response,
    graphql_query_request_factory,
    graphql_response_headers,
    assert_json_response_equals_snapshot,
):
    request = graphql_query_request_factory(query={"test": "error"})
    result = middleware(request, start_response)
    start_response.assert_called_once_with(
        HTTP_STATUS_400_BAD_REQUEST, graphql_response_headers
    )
    assert_json_response_equals_snapshot(result)


def test_attempt_execute_query_with_invalid_variables_returns_error_json(
    middleware,
    start_response,
    graphql_query_request_factory,
    graphql_response_headers,
    assert_json_response_equals_snapshot,
):
    request = graphql_query_request_factory(query=complex_query, variables="invalid")
    result = middleware(request, start_response)
    start_response.assert_called_once_with(
        HTTP_STATUS_400_BAD_REQUEST, graphql_response_headers
    )
    assert_json_response_equals_snapshot(result)


def test_attempt_execute_query_with_invalid_operation_name_string_returns_error_json(
    middleware,
    start_response,
    graphql_query_request_factory,
    graphql_response_headers,
    assert_json_response_equals_snapshot,
):
    request = graphql_query_request_factory(
        query=complex_query, variables=variables, operationName="otherOperation"
    )
    result = middleware(request, start_response)
    start_response.assert_called_once_with(HTTP_STATUS_200_OK, graphql_response_headers)
    assert_json_response_equals_snapshot(result)


def test_attempt_execute_query_with_invalid_operation_name_type_returns_error_json(
    middleware,
    start_response,
    graphql_query_request_factory,
    graphql_response_headers,
    assert_json_response_equals_snapshot,
):
    request = graphql_query_request_factory(
        query=complex_query, variables=variables, operationName=[1, 2, 3]
    )
    result = middleware(request, start_response)
    start_response.assert_called_once_with(
        HTTP_STATUS_400_BAD_REQUEST, graphql_response_headers
    )
    assert_json_response_equals_snapshot(result)


def test_query_is_executed_for_multipart_form_request_with_file(
    middleware, snapshot, start_response, graphql_response_headers
):
    data = """
--------------------------cec8e8123c05ba25
Content-Disposition: form-data; name="operations"

{ "query": "mutation ($file: Upload!) { upload(file: $file) }", "variables": { "file": null } }
--------------------------cec8e8123c05ba25
Content-Disposition: form-data; name="map"

{ "0": ["variables.file"] }
--------------------------cec8e8123c05ba25
Content-Disposition: form-data; name="0"; filename="test.txt"
Content-Type: text/plain

test

--------------------------cec8e8123c05ba25--
    """.rstrip()

    request = create_multipart_request(data)
    result = middleware(request, start_response)
    start_response.assert_called_once_with(HTTP_STATUS_200_OK, graphql_response_headers)
    snapshot.assert_match(result)


class CustomExtension(ExtensionSync):
    def resolve(self, next_, obj, info, **kwargs):
        value = next_(obj, info, **kwargs)
        return f"={value}="


def test_middlewares_and_extensions_are_combined_in_correct_order(schema):
    def test_middleware(next_fn, *args, **kwargs):
        value = next_fn(*args, **kwargs)
        return f"*{value}*"

    app = GraphQL(schema, extensions=[CustomExtension], middleware=[test_middleware])
    client = TestClient(app)
    response = client.post("/", json={"query": '{ hello(name: "BOB") }'})
    assert response.json == {"data": {"hello": "=*Hello, BOB!*="}}
