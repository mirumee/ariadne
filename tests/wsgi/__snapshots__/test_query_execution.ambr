# serializer version: 1
# name: test_attempt_execute_anonymous_subscription_over_post_returns_error_json
  dict({
    'errors': list([
      dict({
        'message': 'Subscription operations can only be executed over a WebSocket connection.',
      }),
    ]),
  })
# ---
# name: test_attempt_execute_complex_query_without_variables_returns_error_json
  dict({
    'data': None,
    'errors': list([
      dict({
        'locations': list([
          dict({
            'column': 18,
            'line': 2,
          }),
        ]),
        'message': "Variable '$name' of required type 'String!' was not provided.",
      }),
    ]),
  })
# ---
# name: test_attempt_execute_query_with_invalid_operation_name_string_returns_error_json
  dict({
    'data': None,
    'errors': list([
      dict({
        'message': "Unknown operation named 'otherOperation'.",
      }),
    ]),
  })
# ---
# name: test_attempt_execute_query_with_invalid_operation_name_type_returns_error_json
  dict({
    'errors': list([
      dict({
        'message': '"[1, 2, 3]" is not a valid operation name.',
      }),
    ]),
  })
# ---
# name: test_attempt_execute_query_with_invalid_variables_returns_error_json
  dict({
    'errors': list([
      dict({
        'message': 'Query variables must be a null or an object.',
      }),
    ]),
  })
# ---
# name: test_attempt_execute_query_with_non_string_query_returns_error_json
  dict({
    'errors': list([
      dict({
        'message': 'The query must be a string.',
      }),
    ]),
  })
# ---
# name: test_attempt_execute_query_without_query_entry_returns_error_json
  dict({
    'errors': list([
      dict({
        'message': 'The query must be a string.',
      }),
    ]),
  })
# ---
# name: test_attempt_execute_subscription_over_post_returns_error_json
  dict({
    'errors': list([
      dict({
        'message': "Operation 'Test' is a subscription and can only be executed over a WebSocket connection.",
      }),
    ]),
  })
# ---
# name: test_complex_query_is_executed_for_post_json_request
  dict({
    'data': dict({
      'hello': 'Hello, Bob!',
    }),
  })
# ---
# name: test_complex_query_without_operation_name_executes_successfully
  dict({
    'data': dict({
      'hello': 'Hello, Bob!',
    }),
  })
# ---
# name: test_query_is_executed_for_multipart_form_request_with_file
  list([
    b'{"data": {"upload": "File"}}',
  ])
# ---
# name: test_query_is_executed_for_post_json_request
  dict({
    'data': dict({
      'status': True,
    }),
  })
# ---
